Create the `Decimal` type

rofrankel opened the pull request:
(Note: this PR is for review purposes only; the contents of this PR should ultimately live in the common-components repo which does not yet exist.)

rofrankel commented on common-components/schemas/type/decimal.md:
```diff
@@ -0,0 +1,37 @@
+
+## Schema
+
+A `Decimal` represents decimal numbesr in a form similar to scientific
```
```suggestion
A `Decimal` represents a decimal number in a form similar to scientific
```

lukesneeringer commented on common-components/proto/api/type/decimal.proto:
```diff
@@ -0,0 +1,36 @@
+// 11/8 (1.375)  === {digits: 1375 exponent: -3}
+message Decimal {
+  // The significant digits (a.k.a. significand or mantissa) of the number.
+  int64 digits = 1;
```
Either base or mantissa would be preferred over "digits".

Honestly I think mantissa is the ideal choice. Yes, nobody knows the word, but it's precise and checking Merriam-Webster or Google is not hard.

rofrankel replied:
Yes, people can always look up a definition, but that certainly has worse ergonomics than something more intuitive.


Wikipedia says (emphasis mine):

> The significand[1] (also mantissa[2] or coefficient,[1] sometimes also argument, or ambiguously fraction[3][4][nb 1] or characteristic[5][2]) is part of a number in scientific notation or in floating-point representation, **consisting of its significant digits**.

Since people Googling "mantissa" will likely end up reading that paragraph and going "oh, it means the significant digits", just naming this field "digits" (or "significant_digits") shortcuts that process with the same end result.

I am happy to be outvoted on this, but I'd like to solicit more opinions.

lukesneeringer replied:
`significant_digits` is actually an interesting idea.

I think the real way to decide is to see what common programming language implementations use. I don't actually know, so let's find out really quick (note: I'll use snake case regardless of language conventions)...

- C++: No library found.
- C#: `lo` / `mid` / `high` (mostly non-public)
- Go: `int_part`
- Java: `unscaled_value`
- PHP: `value`
- Python: `digits`
- Rust: `mantissa`
- Ruby: I can't seem to find a public API term for this in Ruby BigDecimal. Unfortunately, they _do_ use digits, but not as the actual digits of the number, but rather as the _number of significant digits_ (so, e.g. for 3.14159, Ruby `digits` would be 6).

Ooof. _This_ hurts.

lukesneeringer replied:
I think I'm persuaded that `digits` is a bad idea (the Ruby use scares me). But I could be easily persuaded to something else like `unscaled_value`.

I also now oppose `significant_digits` for the Ruby reason, too.

lukesneeringer replied:
Maybe `significand`?

rofrankel replied:
Yeah, I found something similar - there's no consistent precedent set by major languages, which is why I decided to propose something from first principles.

I don't think `significand` is any better or worse than `mantissa`; I'm currently more or less neutral between the two.

I think you have a totally fair and reasonable perspective.  I am going to solicit other opinions in the weekly meeting tomorrow though.

rofrankel replied:
Thanks for the feedback; changed.

lukesneeringer commented on common-components/proto/api/type/decimal.proto:
```diff
@@ -0,0 +1,36 @@
+  // (and implicit leading zeroes, as needed) come after the decmial point.
+  //
+  // When unset, this value is implicitly zero (0).
+  int64 exponent = 2;
```
This isn't really a "pure" exponent. It's a power of 10. I think "e" is the least bad term (with an appropriate comment). The rust_decimal crate uses "scale".

rofrankel replied:
From https://en.wikipedia.org/wiki/Scientific_notation:

> In scientific notation, nonzero numbers are written in the form
>
> *m × 10n*
>
> or m times ten raised to the power of n, where n is an integer, and the coefficient m is a nonzero real number (usually between 1 and 10 in absolute value, and nearly always written as a terminating decimal). The integer n is called the exponent...


(This is more relevant to the question of "digits" vs. "mantissa", but in the interest of completeness and not seeming like I'm cutting off the quote at a cherrypicked point, that section goes on to talk about usage of "significand and mantissa" for *m*, and then goes on to say that "mantissa" is ambiguous.)

rofrankel replied:
Also, it looks like in both Rust and Java, `scale` is the inverse of exponent (i.e. an exponent of 2 is equivalent to a scale of -2).  Do I have that right?

lukesneeringer replied:
> but in the interest of completeness and not seeming like I'm cutting off the quote at a cherrypicked point

Nah, you're fine, and this is persuasive.

> Also, it looks like in both Rust and Java, scale is the inverse of exponent (i.e. an exponent of 2 is equivalent to a scale of -2). Do I have that right?

Yes.

lukesneeringer replied:
I find this persuasive.

toumorokoshi commented on common-components/schemas/type/decimal.md:
```diff
@@ -0,0 +1,37 @@
+  digits (and implicit leading zeroes, as needed) come after the decmial point.
+
+The general formula for converting a `Decimal` to its numeric value is
+`digits * 10^exponent`, where `*` represents multplication and `^` represenets
```
```suggestion
`digits * 10^exponent`, where `*` represents multplication and `^` represents
```

toumorokoshi commented on common-components/schemas/type/decimal.md:
```diff
@@ -0,0 +1,37 @@
+  after the digits.
+
+  When the exponent is less than 0, represents how many of the significant
+  digits (and implicit leading zeroes, as needed) come after the decmial point.
```
```suggestion
  digits (and implicit leading zeroes, as needed) come after the decimal point.
```

toumorokoshi commented on common-components/schemas/type/decimal.md:
```diff
@@ -0,0 +1,37 @@
+  When this value is unset or `0`; in that case, the value of the `Decimal` is
+  simply the value of `digits`.
+
+  When the exponent is greater than 0, represents the number of trailing zeroes
```
```suggestion
  When the exponent is greater than 0, `exponent` represents the number of trailing zeroes
```

rofrankel replied:
Is the way I originally wrote this not idiomatic, i.e. implicitly "[this field] represents"?

toumorokoshi reviewed (approved):
A couple typos I'd like fixed but for sake of speed I'll approve trusting you'll address them. Thank you!

toumorokoshi reviewed (changes requested):
sorry, requesting changes to ensure we switch from `digits` to `significand`.

I might add a "rationale" section to explain why we chose these names, since we're often asked about why we made specific decisions.

rofrankel commented:
Open questions from steering committee meeting:

1. Is it okay to use 64-bit integers at all?  What does that mean for for OAS/JSON? (@hudlow)
1. Should we limit the valid range of a Decimal to that of [decimal64](https://en.wikipedia.org/wiki/Decimal64_floating-point_format)?
  a. Unless my math is wrong, decimal64 doesn't help with the JSON/double problem, because `10^16 - 1` (decimal64 significand range) is greater than `2^53-1`.

rofrankel commented:
> 1. Is it okay to use 64-bit integers at all?  What does that mean for for OAS/JSON? (@hudlow)

My opinion on this is:

* In general, the plan is that each type should have an IDL-agnostic definition, and then (when applicable) reference implementations in IDLs like protobuf and JSON Schema.
* We also expect, in many cases (e.g. money), libraries to be needed in order to make various API implementations of these types (protobuf, JSON Schema, ...) useful in a given language.  That might mean converting them to/from the appropriate in-language representation of a decimal number, or it might mean wrapping them in a class that provides e.g. arithmetic functionality.
* It's consistent with the above to have the following:
  * A JSON Schema for Decimal with _any_ internal implementation details.
  * Libraries in various languages that take JSON objects conforming to that schema and either convert or wrap them.
* If we have those things, then it doesn't actually matter what the JSON wire format is, i.e. it doesn't matter if we encode a numeric field as a string.  The primitive data types supported in JSON aren't actually relevant.

Separate from that line of reasoning:  outside of JSON and ECMAScript, 64-integers are very common.  If we say that the significand of a Decimal is a 53-bit integer just so it can be a JSON number on the  wire, we're making life harder for all the languages which support int64.

And yet another objection:  JSON number (i.e. double) is not a fully correct representation for a 53-bit integer, even if it is closer semantically than a string.  In both cases, the primitive data type supports values that would be invalid, meaning that validation is required.  No language I know of has a 53-bit integer type.

lukesneeringer commented:
> > 2. Should we limit the valid range of a Decimal to that of [decimal64](https://en.wikipedia.org/wiki/Decimal64_floating-point_format)? (@Alfus)
> 
> My opinion on this is:
> 
> * The standard is 15 years old.
> * I cannot find any evidence of support in any upcoming CPU architecture, nor in any programming language.
> * Therefore, it's not clear what practical benefit there would be from hewing to this standard's range limitations.  The cost (in terms of cognitive load, validation, and limited range) doesn't seem to be justified.

I agree with this assessment.

toumorokoshi reviewed (approved):
Thanks! I think I'm happy with it.

I understand the concerns with not using decimal64. But given the lack of adoption after ~15 years I think it's a long shot to recommend the design for that reason.

I think this type does a good job describing the structure, and we can always provide converters to other representations (which, based on Luke's investigation, it looks like we'll have to).

rofrankel commented:
I've updated this PR as discussed in the committee meeting:

1.  Converted the JSON schema to a YAML file.
2.  Added a note about the range exceeding that of double and decimal64.  (@Alfus, if you want something more elaborate, please feel free to suggest and commit a change directly.)

rofrankel commented on common-components/json_schema/decimal.yaml:
```diff
@@ -0,0 +1,38 @@
+      When the exponent is > 0, represents the number of trailing zeroes after
+      the significant digits.
+
+      When the exponent is < 0, represents how many of the significant digits
+      (and implicit leading zeroes, as needed) come after the decmial point.
+
+      When unset, this value is implicitly zero (0).
```
What do you mean by "non-canonical forms" in this context?

rofrankel replied:
@hudlow FYI, we discussed this in your absence today, and decided that we probably want equivalent decimal values to be equal to each other, i.e. it's unfortunate if we endorse both `{significand: 1}` and `{significand: 1, exponent: 0}` as having equal values, but then in practice they're non-equal objects.

So I removed the defaults, as well as references to them.
